x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Authenticator.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authc;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationToken;
+import org.elasticsearch.xpack.core.security.authc.Realm;
+import org.elasticsearch.xpack.core.security.user.User;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public interface Authenticator {
+
+    String name();
+
+    AuthenticationToken extractCredentials(Context context);
+
+    /**
+     * Whether authentication can proceed to authenticate as anonymous or fallback user
+     * if this authenticator can extract a credentials.
+     */
+    default boolean canFollowedByNullTokenHandler() {
+        return true;
+    }
+
+    void authenticate(Context context, ActionListener<Result> listener);
+
+    /**
+     * Gets the token from the <code>Authorization</code> header if the header begins with
+     * <code>Bearer </code>
+     */
+    static SecureString extractBearerTokenFromHeader(ThreadContext threadContext) {
```

Comment:
I added the new method as you suggested and used it for ApiKey (but not `Basic` because it has some difference in error handling which we can look into later). Thanks!